README.md
# <a id="docs" href="#docs">jPanelMenu</a>

### Version 1.4.1

jPanelMenu is a [jQuery](http://jquery.com) plugin for easily creating and managing off-canvas content.

Check out the [demo (and documentation) site](http://jpanelmenu.com) to see it in action.

Check out the [changelog](#changelog) to see what&rsquo;s new.



# <a id="usage" href="#usage">How Do I Use This Thing?</a>

Start off by including the jPanelMenu.js file in your page. (Bonus points for using the minified version [jPanelMenu.min.js], or for bundling the jPanelMenu code into your own JavaScript file to reduce size and HTTP requests.)

Build your page as you normally would (the source order does not matter), and instantiate jPanelMenu by calling the plugin constructor function.

	var jPM = $.jPanelMenu();

By default, jPanelMenu will look for an element with an ID of `menu` to use as the menu, and elements with a class of `menu-trigger` to use as the trigger(s). Either use these IDs and classes on your elements, or pass a custom selector string pointing jPanelMenu to your menu and trigger elements in an object into the constructor function call, as follows:

	var jPM = $.jPanelMenu({
		menu: '#custom-menu-selector',
		trigger: '.custom-menu-trigger-selector'
	});

**Note:** Check out the [options section]($options) for more customizable goodness like the above.

After jPanelMenu has been instantiated (make sure to save the returned object to a variable, as shown above), it&rsquo;s time to turn it on!

	jPM.on();

After that, jPanelMenu will be functioning, and that&rsquo;s it!

If you want to take things to the next level, keep reading.



# <a id="inner-workings" href="#inner-workings">How Does This Thing Work?</a>

When jPanelMenu is [turned on](#api-on), two `<div>` elements are created. The menu element (with an ID of `jPanelMenu-menu`), and the panel element (with a class of `jPanelMenu-panel`). In addition, a class of `jPanelMenu` is applied to the `<html>` tag.

The menu, `#jPanelMenu-menu`, contains the elements targeted by the menu selector passed into the jPanelMenu constructor function. By default, the targeted menu element is cloned into `#jPanelMenu-menu`, and is not removed from its original position in the DOM. This action can be overridden with the [`clone`](#options-clone) option.

The panel, `.jPanelMenu-panel`, contains all of the content in the element specified by the [`panel`](#options-panel) option (except for the elements specified by the [`excludedPanelContent`](#options-excludedPanelContent) option). The selected content is moved, not cloned, into `.jPanelMenu-panel`.

To style or select the menu, use the following selector: `#jPanelMenu-menu`.

To style or select the content panel, use the following selector: `.jPanelMenu-panel`.

When jPanelMenu is [turned off](#api-off), the two `<div>` elements are removed, all of the content inside `.jPanelMenu-panel` is moved back into the `<body>` element, and the class of `jPanelMenu` is removed from the `<html>` tag.



# <a id="animation" href="#animation">Does It Animate?</a>

Of course! (If you want it to, there&rsquo;s an [option](#options) for that.)

Animation is handled by CSS transitions, for browsers with support. CSS transitions are hardware-accelerated on supporting devices, so the animations are silky smooth.

For browsers that do not support CSS transitions, the jQuery animation engine is used as a fallback.



# <a id="options" href="#options">Options</a>

The following options are set via an `object` passed into the constructor function call, as shown below.

	var jPM = $.jPanelMenu({
		menu: '#menu',
		trigger: '.menu-trigger',
		duration: 300
	});

***

### <a href="#options-menu" id="options-menu">menu</a>

A selector string pointing to the desired menu element.

- **Data Type:** `string`
- **Default Value:** `#menu`

***

### <a href="#options-panel" id="options-panel">panel</a>

A selector string pointing to the desired root panel element. Point this to the element containing all content that should go into the panel.

- **Data Type:** `string`
- **Default Value:** `body`

***

### <a href="#options-trigger" id="options-trigger">trigger</a>

A selector string pointing to the menu-triggering element.

- **Data Type:** `string`
- **Default Value:** `.menu-trigger`

***

### <a href="#options-excludedPanelContent" id="options-excludedPanelContent">excludedPanelContent</a>

A selector string specifying which elements within the `<body>` element should **not** be pushed into `.jPanelMenu-panel`. The selector string may contain any selector, not just tags.

Generally, `<style>` and `<script>` tags should not be moved from their original location, but in certain circumstances (mostly advertising), `<script>` tags may need to move with the page content.

- **Data Type:** `string`
- **Default Value:** `style, script`

***

### <a href="#options-clone" id="options-clone">clone</a>

A boolean value specifying whether or not the targeted menu element should be cloned to create `#jPanelMenu-menu`, or simply moved in the DOM.

- **Data Type:** `boolean`
- **Accepted Values:** `true` or `false`
- **Default Value:** `true`

***

### <a href="#options-direction" id="options-direction">direction</a>

A string specifying which direction the menu should open from.

- **Data Type:** `string`
- **Accepted Values:** `left` or `right`
- **Default Value:** `left`

***

### <a href="#options-openPosition" id="options-openPosition">openPosition</a>

The measurement value for the open position of the menu. Can be set as a pixel, percentage, or `em` value.

- **Data Type:** `string`
- **Examples:** `250px`, `75%`, `20em`
- **Default Value:** `250px`

***

### <a href="#options-animated" id="options-animated">animated</a>

A boolean value specifying whether or not the opening and closing of the menu should be animated.

When using the API functions [`open( )`](#api-open), [`close()`](#api-close), and [`trigger()`](#api-trigger), this setting can be overridden by passing in `true` as the parameter. More info in the [API section](#api).

- **Data Type:** `boolean`
- **Accepted Values:** `true` or `false`
- **Default Value:** `true`

***

### <a href="#options-closeOnContentClick" id="options-closeOnContentClick">closeOnContentClick</a>

A boolean value specifying whether or not the menu should be closed when clicking on the panel content.

- **Data Type:** `boolean`
- **Accepted Values:** `true` or `false`
- **Default Value:** `true`

***

### <a href="#options-keyboardShortcuts" id="options-keyboardShortcuts">keyboardShortcuts</a>

An option that allows you to control if keyboard shortcuts are enabled, and if they are, which keys do what.

Setting this option to `false` will disable keyboard shortcuts entirely. To enable keyboard shortcuts, pass in an `array` of `objects`. Each enabled key gets its own `object` in the `array` and each object should be structured as follows:

	{
		code: 27, /* Keycode of enabled key */
		open: true /* Boolean (true or false), specifying whether or not key should open the menu */
		close: false /* Boolean (true or false), specifying whether or not key should close the menu */
	}

- **Data Type:** `array` or `boolean`
- **Accepted Values:** `array` or `false`
- **Default Value:**

	[
		{
			code: 27, /* Escape Key */
			open: false,
			close: true 
		},{
			code: 37, /* Left Arrow Key */
			open: false,
			close: true 
		},{
			code: 39, /* Right Arrow Key */
			open: true,
			close: true 
		},{
			code: 77, /* M Key */
			open: true,
			close: true 
		}
	]

***

### <a href="#options-duration" id="options-duration">duration</a>

The time, in milliseconds, which it should take to open and close the menu, when animated.

- **Data Type:** `int`
- **Default Value:** `150`

***

### <a href="#options-openDuration" id="options-openDuration">openDuration</a>

The time, in milliseconds, which it should take to open the menu, when animated. If set, this overrides the duration option.

- **Data Type:** `int`
- **Default Value:** Inherited from [`duration`](#options-duration)

***

### <a href="#options-closeDuration">closeDuration</a>

The time, in milliseconds, which it should take to close the menu, when animated. If set, this overrides the duration option.

- **Data Type:** `int`
- **Default Value:** Inherited from [`duration`]()

***

### <a href="#options-easing" id="options-easing">easing</a>

The easing function to use when animating the opening and closing of the menu.

- **Data Type:** `string`
- **Accepted Values:** `linear`, `ease`, `ease-in`, `ease-out`, `ase-in-out`
- **Default Value:** `ease-in-out`

***

### <a href="#options-openEasing" id="options-openEasing">openEasing</a>

The easing function to use when animating the opening of the menu. If set, this overrides the easing option.

- **Data Type:** `string`
- **Accepted Values:** `linear`, `ease`, `ease-in`, `ease-out`, `ase-in-out`
- **Default Value:** Inherited from [`easing`](#options-easing)

***

### <a href="#options-closeEasing" id="options-closeEasing">closeEasing</a>

The easing function to use when animating the closing of the menu. If set, this overrides the easing option.

- **Data Type:** `string`
- **Accepted Values:** `linear`, `ease`, `ease-in`, `ease-out`, `ase-in-out`
- **Default Value:** Inherited from [`easing`](#options-easing)

***

### <a href="#options-before" id="options-before">before</a>

Called before the menu is opened or closed, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-beforeOpen" id="options-beforeOpen">beforeOpen</a>

Called before the menu is opened, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-beforeClose" id="options-beforeClose">beforeClose</a>

Called before the menu is closed, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-after" id="options-after">after</a>

Called after the menu is opened or closed, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-afterOpen" id="options-afterOpen">afterOpen</a>

Called after the menu is opened, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-afterClose" id="options-afterClose">afterClose</a>

Called after the menu is closed, regardless of animation state.

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-beforeOn" id="options-beforeOn">beforeOn</a>

Called before the plugin is turned on (when [`on( )`](#api-on) is called).

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-afterOn" id="options-afterOn">afterOn</a>

Called after the plugin is turned on (when [`on( )`](#api-on) is called).

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-beforeOff" id="options-beforeOff">beforeOff</a>

Called before the plugin is turned off (when [`off( )`](#api-off) is called).

- **Data Type:** `function`
- **Default Value:** `function(){ }`

***

### <a href="#options-afterOff" id="options-afterOff">afterOff</a>

Called after the plugin is turned off (when [`off( )`](#api-off) is called).

- **Data Type:** `function`
- **Default Value:** `function(){ }`

# <a id="api" href="#api">API</a>

The following are the methods and properties of the object returned by the jPanelMenu constructor function call. In the following example, these would be the methods and properties of `jPM`.

	var jPM = $.jPanelMenu();

	jPM.on();

	jPM.trigger(true);

### <a href="#api-on" id="api-on">on(&nbsp;)</a>

Initializes a jPanelMenu instance. Sets up the markup, styles, listeners, and interactions, according to the options passed into the constructor function.

- **Returns:** `null`

***

### <a href="#api-off" id="api-off">off(&nbsp;)</a>

Destroys a jPanelMenu instance. Resets the markup and styles, removes listeners and interactions.

- **Returns:** `null`

***

### <a href="#api-trigger" id="api-trigger">trigger(&nbsp;`animated`&nbsp;)</a>

Triggers the opening or closing of the menu, depending on the current state (open or closed).

- **Parameters:**
	- `animated`
		- A boolean value that determines whether or not to animate the action. The action will animate if set to `true`, and will not animate if set to `false`. If no value is set, the value of the [`animated`](#options-animated) option will be used.
		- **Data Type:** `boolean`
		- **Accepted Values:** `true` or `false`
- **Returns:** `null`

***

### <a href="#api-open" id="api-open">open(&nbsp;`animated`&nbsp;)</a>

Triggers the opening of the menu.

- **Parameters:**
	- `animated`
		- A boolean value that determines whether or not to animate the action. The action will animate if set to `true`, and will not animate if set to `false`. If no value is set, the value of the [`animated`](#options-animated) option will be used.
		- **Data Type:** `boolean`
		- **Accepted Values:** `true` or `false`
- **Returns:** `null`

***

### <a href="#api-close" id="api-close">close(&nbsp;`animated`&nbsp;)</a>

Triggers the closing of the menu.

- **Parameters:**
	- `animated`
		- A boolean value that determines whether or not to animate the action. The action will animate if set to `true`, and will not animate if set to `false`. If no value is set, the value of the [`animated`](#options-animated) option will be used.
		- **Data Type:** `boolean`
		- **Accepted Values:** `true` or `false`
- **Returns:** `null`

***

###<a href="#api-isOpen" id="api-isOpen">isOpen(&nbsp;)</a>

Checks the current state of the menu. Returns `true` if the menu is currently open, and `false` if it is closed.

- **Returns:** `boolean`, `true` or `false`

***

###<a href="#api-menu" id="api-menu">menu</a>

A property equal to the raw selector string of the created menu object.

- **Data Type:** `string`

***

### <a href="#api-getMenu" id="api-getMenu">getMenu(&nbsp;)</a>

Returns a `jQuery Object` containing the created menu object.

- **Returns:** `jQuery Object`

***

### <a href="#api-panel" id="api-panel">panel</a>

A property equal to the raw selector string of the created panel object.

- **Data Type:** `string`

***

### <a href="#api-getPanel" id="api-getPanel">getPanel(&nbsp;)</a>

Returns a `jQuery Object` containing the created panel object.

- **Returns:** `jQuery Object`

***


changelog.md
# 1.4.1

### November 11th, 2014

- [ADDED] `touchend` listeners for better touch support.

# 1.4.0

### November 11th, 2014

- [ADDED] [`panel`](#options-panel) option.
- [ADDED] [`clone`](#options-clone) option.
- [ADDED] [`setPosition(&nbsp;)`](#api-setPosition) API method.
- [REMOVED] Support for fixed positioning within the panel. CSS transforms and fixed positioning [do not get along well, per the spec](http://meyerweb.com/eric/thoughts/2011/09/12/un-fixing-fixed-elements-with-css-transforms/). If fixed positioning is needed, use [the legacy build](https://github.com/acolangelo/jPanelMenu/blob/master/jquery.jpanelmenu-legacy.js) in the jPanelMenu repository.
- [UPDATED] `.jPanelMenu-panel` to be positioned statically.
- [UPDATED] Background handling so that all properties are transferred to the `.jPanelMenu-panel` appropriately.
- [UPDATED] Key press preventers to include typing within a `<select>` field.
- [FIXED] Event propagation up to the document.
- [FIXED] An issue causing links under the menu button to be triggered inadvertently.
- [FIXED] An issue with loop styles and the Ember.js framework.

# 1.3.0

### February 4th, 2013

- [ADDED] [`closeOnContentClick`](#options-closeOnContentClick) option.

# 1.2.0

### February 3rd, 2013

- [ADDED] New callback options: [`beforeOn`](#options-beforeOn), [`afterOn`](#options-afterOn), [`beforeOff`](#options-beforeOff), [`afterOff`](#options-afterOff)

# 1.1.1

### February 3rd, 2013

- [FIXED] A conflict between [keyboard shortcuts](#options-keyboardShortcuts) and text inputs. (Thanks to [stoeffel](https://github.com/stoeffel).)
- Renamed JavaScript resources to be more friendly for future development.

# 1.1.0

### December 7th, 2012

- [ADDED] [Directional control](#options-direction). Panel can now slide left or right.
- [REMOVED] Unnecessary trigger check condition. (Thanks to [James Wilson](https://twitter.com/jimmynotjim)!)

# 1.0.0

### November 4th, 2012

- First release of jPanelMenu.


jPanelMenu-1.0.0.js
/**
  *
  * jPanelMenu 1.0.0 (http://jpanelmenu.com)
  * By Anthony Colangelo (http://acolangelo.com)
  *
* */

(function($){
	$.jPanelMenu = function(options) {
		if ( typeof(options) == "undefined" || options == null ) { options = {}; };

		var jP = {
			options: $.extend({
				menu: '#menu',
				trigger: '.menu-trigger',
				excludedPanelContent: 'style, script',

				openPosition: '250px',
				animated: true,

				keyboardShortcuts: [
					{
						code: 27,
						open: false,
						close: true
					},
					{
						code: 37,
						open: false,
						close: true
					},
					{
						code: 39,
						open: true,
						close: true
					},
					{
						code: 77,
						open: true,
						close: true
					}
				],

				duration: 150,
				openDuration: options.duration || 150,
				closeDuration: options.duration || 150,

				easing: 'ease-in-out',
				openEasing: options.easing || 'ease-in-out',
				closeEasing: options.easing || 'ease-in-out',

				before: function(){ },
				beforeOpen: function(){ },
				beforeClose: function(){ },

				after: function(){ },
				afterOpen: function(){ },
				afterClose: function(){ }
			},options),

			settings: {
				transitionsSupported:	'WebkitTransition' in document.body.style ||
										'MozTransition' in document.body.style ||
										'msTransition' in document.body.style ||
										'OTransition' in document.body.style ||
										'Transition' in document.body.style
				,
				shiftFixedChildren: false,
				panelPosition: 'relative',
				positionUnits: 'px'
			},

			menu: '#jPanelMenu-menu',

			panel: '.jPanelMenu-panel',

			fixedChildren: [],

			timeouts: {},

			clearTimeouts: function() {
				clearTimeout(jP.timeouts.open);
				clearTimeout(jP.timeouts.afterOpen);
				clearTimeout(jP.timeouts.afterClose);
			},

			setPositionUnits: function() {
				var foundUnit = false,
					allowedUnits = ['%','px','em']
				;

				for ( unitID in allowedUnits ) {
					var unit = allowedUnits[unitID];
					if ( jP.options.openPosition.toString().substr(-unit.length) == unit )
					{
						foundUnit = true;
						jP.settings.positionUnits = unit;
					}
				}

				if ( !foundUnit ) { jP.options.openPosition = parseInt(jP.options.openPosition) + jP.settings.positionUnits }
			},

			checkFixedChildren: function() {
				jP.disableTransitions();

				var defaultPanelStyle = {
					position: $(jP.panel).css('position'),
					left: ($(jP.panel).css('left') == 'auto')?0:$(jP.panel).css('left')
				};

				$(jP.panel).find('> *').each(function(){
					if ( $(this).css('position') == 'fixed' && $(this).css('left') == 'auto' ) { jP.fixedChildren.push(this); }
				});

				if ( jP.fixedChildren.length > 0 )
				{
					jP.setPanelStyle({
						position: 'relative',
						left: '1px'
					});

					if ( parseInt($(jP.fixedChildren[0]).offset().left) == 0 ) { jP.settings.shiftFixedChildren = true; }
				}

				jP.setPanelStyle(defaultPanelStyle);
			},

			setjPanelMenuStyles: function() {
				var bgColor = '#fff';
				var htmlBG = $('html').css('background-color');
				var bodyBG = $('body').css('background-color');
				var vpHeight = $(window).height();

				if ( bodyBG != 'transparent' && bodyBG != "rgba(0, 0, 0, 0)") { bgColor = bodyBG; }
				else if ( htmlBG != 'transparent' && htmlBG != "rgba(0, 0, 0, 0)") { bgColor = htmlBG; }
				else { bgColor = '#fff'; }

				if ( $('#jPanelMenu-style-master').length == 0 )
				{
					$('body').append('<style id="jPanelMenu-style-master">body{width:100%}.jPanelMenu,body{overflow-x:hidden}#jPanelMenu-menu{display:block;position:fixed;top:0;left:0;height:100%;z-index:-1;overflow-x:hidden;overflow-y:scroll;-webkit-overflow-scrolling:touch}.jPanelMenu-panel{position:static;left:0;top:0;z-index:2;width:100%;min-height:' + vpHeight + 'px;background:' + bgColor + '}</style>');
				}
			},

			setMenuState: function(open) {
				var position = (open)?'open':'closed';
				$('body').attr('data-menu-position', position);
			},

			getMenuState: function() {
				return $('body').attr('data-menu-position');
			},

			menuIsOpen: function() {
				if ( jP.getMenuState() == 'open' ) return true;
				else return false;
			},

			setMenuStyle: function(styles) {
				$(jP.menu).css(styles);
			},

			setPanelStyle: function(styles) {
				$(jP.panel).css(styles);
			},

			showMenu: function() {
				jP.setMenuStyle({
					display: 'block'
				});
				jP.setMenuStyle({
					'z-index': '1'
				});
			},

			hideMenu: function() {
				jP.setMenuStyle({
					'z-index': '-1'
				});
				jP.setMenuStyle({
					display: 'none'
				});
			},

			getPrefix: function () {
				if('result' in arguments.callee) return arguments.callee.result;
				var regex = /^(Moz|Webkit|Khtml|O|ms|Icab)(?=[A-Z])/;
				var someScript = document.getElementsByTagName('script')[0];
				for(var prop in someScript.style)
				{
					if(regex.test(prop))
					{
					   var prefix = prop.match(regex)[0];
					   return arguments.callee.result = '-' + prefix.toLowerCase() + '-';
					}
				}
				if('WebkitOpacity' in someScript.style) return arguments.callee.result = '-webkit-';
				if('KhtmlOpacity' in someScript.style) return arguments.callee.result = '-khtml-';
				return arguments.callee.result = '';
			},

			enableTransitions: function(duration, easing) {
				var formattedDuration = duration/1000;
				var formattedEasing = jP.getCSSEasingFunction(easing);
				var formattedTransition = jP.getPrefix('transition') + 'transition';
				jP.disableTransitions();
				$('body').append('<style id="jPanelMenu-style-transitions">.jPanelMenu-panel{' + formattedTransition + ': all ' + formattedDuration + 's ' + formattedEasing + ';}</style>');
			},

			disableTransitions: function() {
				$('#jPanelMenu-style-transitions').remove();
			},

			enableFixedTransitions: function(selector, id, duration, easing) {
				var formattedDuration = duration/1000;
				var formattedEasing = jP.getCSSEasingFunction(easing);
				var formattedTransition = jP.getPrefix('transition') + 'transition';
				jP.disableFixedTransitions(id);
				$('body').append('<style id="jPanelMenu-style-fixed-' + id + '">' + selector + '{' + formattedTransition + ': all ' + formattedDuration + 's ' + formattedEasing + ';}</style>');
			},

			disableFixedTransitions: function(id) {
				$('#jPanelMenu-style-fixed-' + id).remove();
			},

			getCSSEasingFunction: function(name) {
				switch ( name )
				{
					case 'linear':
						return name;
						break;

					case 'ease':
						return name;
						break;

					case 'ease-in':
						return name;
						break;

					case 'ease-out':
						return name;
						break;

					case 'ease-in-out':
						return name;
						break;

					default:
						return 'ease-in-out';
						break;
				}
			},

			getJSEasingFunction: function(name) {
				switch ( name )
				{
					case 'linear':
						return name;
						break;

					default:
						return 'swing';
						break;
				}
			},

			testTransform: function() {
				var el = document.createElement('p'),
				has3d,
				transforms = {
					'WebkitTransform':'-webkit-transform',
					'OTransform':'-o-transform',
					'MSTransform':'-ms-transform',
					'MozTransform':'-moz-transform',
					'Transform':'transform'
				};

				document.body.insertBefore(el, null);

				for(var t in transforms){
					if( el.style[t] !== undefined ){
						el.style[t] = 'translate3d(1px,1px,1px)';
						has3d = window.getComputedStyle(el).getPropertyValue(transforms[t]);
					}
				}

				document.body.removeChild(el);

				if (has3d !== undefined && has3d.length > 0 && has3d !== "none") return true;
				else return false;
			},

			openMenu: function(animated) {
				if ( typeof(animated) == "undefined" || animated == null ) { animated = jP.options.animated };

				jP.clearTimeouts();

				jP.options.before();
				jP.options.beforeOpen();

				jP.setMenuState(true);

				jP.setPanelStyle({ position: 'relative' });

				jP.showMenu();

				var animationChecks = {
					none: (!animated)?true:false,
					transitions: (animated && jP.settings.transitionsSupported)?true:false
				};

				if ( animationChecks.transitions || animationChecks.none ) {
					if ( animationChecks.none ) jP.disableTransitions();
					if ( animationChecks.transitions ) jP.enableTransitions(jP.options.openDuration, jP.options.openEasing);

					if(jP.testTransform) {
						var transform = jP.getPrefix('transform') + 'transform';
						$(jP.panel).css(transform,'translate3d(' + jP.options.openPosition + ',0,0)');
					}
					else {

						jP.setPanelStyle({ left: jP.options.openPosition });

						if ( jP.settings.shiftFixedChildren )
						{
							$(jP.fixedChildren).each(function(){
								var id = $(this).prop("tagName").toLowerCase() + ' ' + $(this).attr('class'),
									selector = id.replace(' ','.'),
									id = id.replace(' ','-')
								;

								if ( animationChecks.none ) jP.disableFixedTransitions(id);
								if ( animationChecks.transitions ) jP.enableFixedTransitions(selector, id, jP.options.openDuration, jP.options.openEasing);

								$(this).css({
									left: jP.options.openPosition
								});
							});
						}
					}

					jP.timeouts.afterOpen = setTimeout(function(){
						jP.disableTransitions();
						if ( jP.settings.shiftFixedChildren )
						{
							$(jP.fixedChildren).each(function(){
								var id = $(this).prop("tagName").toLowerCase() + ' ' + $(this).attr('class'),
									id = id.replace(' ','-')
								;

								jP.disableFixedTransitions(id);
							});
						}

						jP.options.after();
						jP.options.afterOpen();
						jP.initiateContentTouchListeners();
						jP.initiateContentClickListeners();
					}, jP.options.openDuration);
				}
				else {
					var formattedEasing = jP.getJSEasingFunction(jP.options.openEasing);

					$(jP.panel).stop().animate({
						left: jP.options.openPosition
					}, jP.options.openDuration, formattedEasing, function(){
						jP.options.after();
						jP.options.afterOpen();
						jP.initiateContentTouchListeners();
						jP.initiateContentClickListeners();
					});

					if ( jP.settings.shiftFixedChildren )
					{
						$(jP.fixedChildren).each(function(){
							$(this).stop().animate({
								left: jP.options.openPosition
							}, jP.options.openDuration, formattedEasing);
						});
					}
				}
			},

			closeMenu: function(animated) {
				if ( typeof(animated) == "undefined" || animated == null ) { animated = jP.options.animated };

				jP.clearTimeouts();

				jP.options.before();
				jP.options.beforeClose();

				jP.setMenuState(false);

				var animationChecks = {
					none: (!animated)?true:false,
					transitions: (animated && jP.settings.transitionsSupported)?true:false
				};

				if ( animationChecks.transitions || animationChecks.none ) {
					if ( animationChecks.none ) jP.disableTransitions();
					if ( animationChecks.transitions ) jP.enableTransitions(jP.options.closeDuration, jP.options.closeEasing);

					if(jP.testTransform) {
						transform = jP.getPrefix('transform') + 'transform';
						$(jP.panel).css(transform,'translate3d(0,0,0)');
					}
					else {
						jP.setPanelStyle({ left: 0 + jP.settings.positionUnits });

						if ( jP.settings.shiftFixedChildren )
						{
							$(jP.fixedChildren).each(function(){
								var id = $(this).prop("tagName").toLowerCase() + ' ' + $(this).attr('class'),
									selector = id.replace(' ','.'),
									id = id.replace(' ','-')
								;

								if ( animationChecks.none ) jP.disableFixedTransitions(id);
								if ( animationChecks.transitions ) jP.enableFixedTransitions(selector, id, jP.options.closeDuration, jP.options.closeEasing);

								$(this).css({
									left: 0 + jP.settings.positionUnits
								});
							});
						}
					}

					jP.timeouts.afterClose = setTimeout(function(){
						jP.setPanelStyle({ position: jP.settings.panelPosition });

						jP.disableTransitions();
						if ( jP.settings.shiftFixedChildren )
						{
							$(jP.fixedChildren).each(function(){
								var id = $(this).prop("tagName").toLowerCase() + ' ' + $(this).attr('class'),
									id = id.replace(' ','-')
								;

								jP.disableFixedTransitions(id);
							});
						}

						jP.hideMenu();
						jP.options.after();
						jP.options.afterClose();
						jP.destroyContentTouchListeners();
						jP.destroyContentClickListeners();
					}, jP.options.closeDuration);
				}
				else {
					var formattedEasing = jP.getJSEasingFunction(jP.options.closeEasing);

					$(jP.panel).stop().animate({
						left: '0%'
					}, jP.options.closeDuration, formattedEasing, function(){
						jP.setPanelStyle({ position: jP.settings.panelPosition });

						jP.hideMenu();
						jP.options.after();
						jP.options.afterClose();
						jP.destroyContentTouchListeners();
						jP.destroyContentClickListeners();
					});

					if ( jP.settings.shiftFixedChildren )
					{
						$(jP.fixedChildren).each(function(){
							$(this).stop().animate({
								left: 0 + jP.settings.positionUnits
							}, jP.options.closeDuration, formattedEasing);
						});
					}
				}
			},

			triggerMenu: function(animated) {
				if ( jP.menuIsOpen() ) jP.closeMenu(animated);
				else jP.openMenu(animated);
			},

			initiateClickListeners: function() {
				$(document).on('click',jP.options.trigger,function(){ jP.triggerMenu(jP.options.animated); return false; });
			},

			destroyClickListeners: function() {
				$(document).off('click',jP.options.trigger,null);
			},

			initiateContentClickListeners: function() {
				$(document).on('click',jP.panel,function(e){
					if ( jP.menuIsOpen() )
					{
						if ( !$(e.target).hasClass('menu-trigger') ) { jP.closeMenu(jP.options.animated); }
					}
				});
			},

			destroyContentClickListeners: function() {
				$(document).off('click',jP.panel,null);
			},

			initiateKeyboardListeners: function() {
				$(document).on('keydown',function(e){
					for ( mapping in jP.options.keyboardShortcuts ) {
						if ( e.which == jP.options.keyboardShortcuts[mapping].code )
						{
							var key = jP.options.keyboardShortcuts[mapping];

							if ( key.open && key.close ) { jP.triggerMenu(jP.options.animated); }
							else if ( (key.open && !key.close) && !jP.menuIsOpen() ) { jP.openMenu(jP.options.animated); }
							else if ( (!key.open && key.close) && jP.menuIsOpen() ) { jP.closeMenu(jP.options.animated); }

							return false;
						}
					}
				});
			},

			destroyKeyboardListeners: function() {
				$(document).off('keydown',null);
			},

			initiateTouchListeners: function() {

				$(document).on('touchstart',jP.options.trigger,function(e){
					var touch = event.touches[0];
					xOrg = touch.pageX;
					xCur = xOrg;

					e.preventDefault();

				});

				$(document).on('touchmove',jP.options.trigger,function(e){
					jP.showMenu();

					var touch = event.touches[0];
					xCur = touch.pageX;
					xDif = (xCur - xOrg);
					offSet = parseInt(jP.options.openPosition);
					half = offSet * '.5';
					if(xDif > 0) {
						xDis = xDif;
						transDis = 'translate3d(' + xDis + 'px,0,0)';
					}
					else {
						xDis = (xDif * -1);
						transDis = 'translate3d(' + (offSet - xDis) + 'px,0,0)';
					}

					if( xCur < offSet && xDif > 0 )
					{
						if(jP.testTransform) {
							transform = jP.getPrefix('transform') + 'transform';
							$(jP.panel).css(transform, transDis);
						}
						else {
							jP.setPanelStyle({ left : xDis +'px'});
						}
					 }
					else if( xDif < 0 && (offSet - xDis) > 0 && jP.menuIsOpen() )
					{
						if(jP.testTransform) {
							transform = jP.getPrefix('transform') + 'transform';
							$(jP.panel).css(transform, transDis);
						}
						else {
							jP.setPanelStyle({ left : (offSet - xDis) + 'px' });
						}
					}

				});

				$(document).on('touchend',jP.options.trigger,function(e){
					xEnd = xCur;

					if ( xEnd === xOrg )
					{
						jP.triggerMenu(jP.options.animated); return false;
					}
					else if ( jP.menuIsOpen() )
					{
						if (xDis > half) jP.closeMenu();
						else jP.openMenu();
					}
					else
					{
						if (xDis > half) jP.openMenu();
						else jP.closeMenu();
					}
				});
			},

			destroyTouchListeners: function() {
				$(document).off('touchstart',jP.options.trigger,null);
				$(document).off('touchmove',jP.options.trigger,null);
				$(document).off('touchend',jP.options.trigger,null);
			},

			initiateContentTouchListeners: function() {

				$(document).on('touchstart',jP.panel,function(e){
					var touch = event.touches[0];
					xOrg = touch.pageX;
					xCur = xOrg;

					if( jP.menuIsOpen() ){ e.preventDefault(); }

				});

				$(document).on('touchmove',jP.panel,function(e){

					var touch = event.touches[0];
					xCur = touch.pageX;
					xDif = (xCur - xOrg);
					offSet = parseInt(jP.options.openPosition);
					half = offSet * '.5';
					if(xDif > 0) {
						xDis = xDif;
						transDis = 'translate3d(' + xDis + 'px,0,0)';
					}
					else {
						xDis = (xDif * -1);
						transDis = 'translate3d(' + (offSet - xDis) + 'px,0,0)';
					}

					if( xDif < 0 && (offSet - xDis) > 0 && jP.menuIsOpen() )
					{
						if(jP.testTransform) {
							transform = jP.getPrefix('transform') + 'transform';
							$(jP.panel).css(transform, transDis);
						}
						else {
							jP.setPanelStyle({ left : (offSet - xDis) + 'px' });
						}
					}

				});

				$(document).on('touchend',jP.panel,function(e){
					xEnd = xCur;

					if ( jP.menuIsOpen() )
					{

						if ( xEnd === xOrg )
						{
							jP.closeMenu(jP.options.animated);
						}
						else
						{
							if (xDis > half) jP.closeMenu();
							else jP.openMenu();
						}
					}
				});

			},

			destroyContentTouchListeners: function() {
				$(document).off('touchstart',jP.panel,null);
				$(document).off('touchmove',jP.panel,null);
				$(document).off('touchend',jP.panel,null);
			},

			setupMarkup: function() {
				$('html').addClass('jPanelMenu');
				$('body > *').not(jP.menu + ', ' + jP.options.excludedPanelContent).wrapAll('<div class="' + jP.panel.replace('.','') + '"/>');
				$(jP.options.menu).clone().attr('id', jP.menu.replace('#','')).insertAfter('body > ' + jP.panel);
			},

			resetMarkup: function() {
				$('html').removeClass('jPanelMenu');
				$('body > ' + jP.panel + ' > *').unwrap();
				$(jP.menu).remove();
			},

			init: function() {
				jP.initiateTouchListeners();
				jP.initiateClickListeners();
				if ( Object.prototype.toString.call(jP.options.keyboardShortcuts) === '[object Array]' ) { jP.initiateKeyboardListeners(); }

				jP.setjPanelMenuStyles();
				jP.setMenuState(false);
				jP.setupMarkup();

				jP.setMenuStyle({
					width: jP.options.openPosition
				});

				jP.checkFixedChildren();
				jP.setPositionUnits();

				jP.closeMenu(false);
			},

			destroy: function() {
				jP.closeMenu();
				jP.destroyTouchListeners();
				jP.destroyClickListeners();
				if ( Object.prototype.toString.call(jP.options.keyboardShortcuts) === '[object Array]' ) { jP.destroyKeyboardListeners(); }

				jP.resetMarkup();
				$(jP.fixedChildren).each(function(){ $(this).css({ left: 'auto' }); });
				jP.fixedChildren = [];
			}
		};

		return {
			on: jP.init,
			off: jP.destroy,
			trigger: jP.triggerMenu,
			open: jP.openMenu,
			close: jP.closeMenu,
			isOpen: jP.menuIsOpen,
			menu: jP.menu,
			getMenu: function() { return $(jP.menu); },
			panel: jP.panel,
			getPanel: function() { return $(jP.panel); }
		};
	};
})(jQuery);
